From 76595b796385c23c2c68f01b9e033cf551a53ea6 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Mon, 11 Jan 2016 15:23:55 +0000
Subject: [PATCH 123/166] security,perf: Allow further restriction of perf_event_open

When kernel.perf_event_open is set to 3 (or greater), disallow all
access to performance events by users without CAP_SYS_ADMIN or
CAP_PERFMON.
Add a Kconfig symbol CONFIG_SECURITY_PERF_EVENTS_RESTRICT that
makes this value the default.

This is based on a similar feature in grsecurity
(CONFIG_GRKERNSEC_PERF_HARDEN).  This version doesn't include making
the variable read-only.  It also allows enabling further restriction
at run-time regardless of whether the default is changed.

As part of the v5.5 linux-hardened rebase, this commit was adapted to
work with the new perf_event LSM hooks, introduced in da97e18458fb42
("perf_event: Add support for LSM and SELinux checks").

As part of the v5.8 linux-hardened rebase, this commit was adapted to
work with the new CAP_PERFMON capability.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
[levente@leventepolyak.net: Adapt to work with the new perf_event LSM hooks]
Signed-off-by: Levente Polyak <levente@leventepolyak.net>
[thibaut.sautereau@ssi.gouv.fr: Adapt to work with the new CAP_PERFMON capability]
Signed-off-by: Thibaut Sautereau <thibaut.sautereau@ssi.gouv.fr>
Signed-off-by: Levente Polyak <levente@leventepolyak.net>
---
 Documentation/admin-guide/sysctl/kernel.rst | 2 ++
 include/linux/perf_event.h                  | 8 ++++++++
 kernel/events/core.c                        | 7 ++++++-
 security/Kconfig                            | 9 +++++++++
 tools/perf/Documentation/security.txt       | 1 +
 tools/perf/util/evsel.c                     | 1 +
 6 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/Documentation/admin-guide/sysctl/kernel.rst b/Documentation/admin-guide/sysctl/kernel.rst
index f3ee807b5d8b..3bcd46bfe95e 100644
--- a/Documentation/admin-guide/sysctl/kernel.rst
+++ b/Documentation/admin-guide/sysctl/kernel.rst
@@ -989,10 +989,12 @@ with respect to CAP_PERFMON use cases.
      Disallow raw tracepoint access by users without ``CAP_PERFMON``.
 
 >=1  Disallow CPU event access by users without ``CAP_PERFMON``.
 
 >=2  Disallow kernel profiling by users without ``CAP_PERFMON``.
+
+>=3  Disallow use of any event by users without ``CAP_PERFMON``.
 ===  ==================================================================
 
 
 perf_event_max_stack
 ====================
diff --git a/include/linux/perf_event.h b/include/linux/perf_event.h
index fd1d91017b99..8c4db071cb56 100644
--- a/include/linux/perf_event.h
+++ b/include/linux/perf_event.h
@@ -1774,10 +1774,18 @@ static inline int perf_is_paranoid(void)
 	return sysctl_perf_event_paranoid > -1;
 }
 
 extern int perf_allow_kernel(void);
 
+static inline int perf_allow_open(void)
+{
+	if (sysctl_perf_event_paranoid > 2 && !perfmon_capable())
+		return -EACCES;
+
+	return security_perf_event_open(PERF_SECURITY_OPEN);
+}
+
 static inline int perf_allow_cpu(void)
 {
 	if (sysctl_perf_event_paranoid > 0 && !perfmon_capable())
 		return -EACCES;
 
diff --git a/kernel/events/core.c b/kernel/events/core.c
index d95f9dce018f..5d48dc9c1c78 100644
--- a/kernel/events/core.c
+++ b/kernel/events/core.c
@@ -461,12 +461,17 @@ static struct kmem_cache *perf_event_cache;
  * perf event paranoia level:
  *  -1 - not paranoid at all
  *   0 - disallow raw tracepoint access for unpriv
  *   1 - disallow cpu events for unpriv
  *   2 - disallow kernel profiling for unpriv
+ *   3 - disallow all unpriv perf event use
  */
+#ifdef CONFIG_SECURITY_PERF_EVENTS_RESTRICT
+int sysctl_perf_event_paranoid __read_mostly = 3;
+#else
 int sysctl_perf_event_paranoid __read_mostly = 2;
+#endif
 
 /* Minimum for 512 kiB + 1 user control page. 'free' kiB per user. */
 static int sysctl_perf_event_mlock __read_mostly = 512 + (PAGE_SIZE / 1024);
 
 /*
@@ -13413,11 +13418,11 @@ SYSCALL_DEFINE5(perf_event_open,
 	err = perf_copy_attr(attr_uptr, &attr);
 	if (err)
 		return err;
 
 	/* Do we allow access to perf_event_open(2) ? */
-	err = security_perf_event_open(PERF_SECURITY_OPEN);
+	err = perf_allow_open();
 	if (err)
 		return err;
 
 	if (!attr.exclude_kernel) {
 		err = perf_allow_kernel();
diff --git a/security/Kconfig b/security/Kconfig
index 132b6fb4bc7e..ea669f5e261b 100644
--- a/security/Kconfig
+++ b/security/Kconfig
@@ -70,10 +70,19 @@ config MSEAL_SYSTEM_MAPPINGS
 	  this config can't be enabled universally.
 
 	  For complete descriptions of memory sealing, please see
 	  Documentation/userspace-api/mseal.rst
 
+config SECURITY_PERF_EVENTS_RESTRICT
+	bool "Restrict unprivileged use of performance events"
+	depends on PERF_EVENTS
+	help
+	  If you say Y here, the kernel.perf_event_paranoid sysctl
+	  will be set to 3 by default, and no unprivileged use of the
+	  perf_event_open syscall will be permitted unless it is
+	  changed.
+
 config SECURITY
 	bool "Enable different security models"
 	depends on SYSFS
 	depends on MULTIUSER
 	default y
diff --git a/tools/perf/Documentation/security.txt b/tools/perf/Documentation/security.txt
index 4fe3b8b1958f..a7d88cc23a70 100644
--- a/tools/perf/Documentation/security.txt
+++ b/tools/perf/Documentation/security.txt
@@ -146,10 +146,11 @@ Perf tool provides a message similar to the one below:
      -1: Allow use of (almost) all events by all users
          Ignore mlock limit after perf_event_mlock_kb without CAP_IPC_LOCK
    >= 0: Disallow raw and ftrace function tracepoint access
    >= 1: Disallow CPU event access
    >= 2: Disallow kernel profiling
+   >= 3: Disallow use of any event
    To make the adjusted perf_event_paranoid setting permanent preserve it
    in /etc/sysctl.conf (e.g. kernel.perf_event_paranoid = <setting>)
 
 To make sure that access is limited by MAC policy settings inspect system
 audit records using journalctl command or /var/log/audit/audit.log so the
diff --git a/tools/perf/util/evsel.c b/tools/perf/util/evsel.c
index 9df9818e3701..eef46803b2cf 100644
--- a/tools/perf/util/evsel.c
+++ b/tools/perf/util/evsel.c
@@ -3848,10 +3848,11 @@ int evsel__open_strerror(struct evsel *evsel, struct target *target,
 		 "  -1: Allow use of (almost) all events by all users\n"
 		 "      Ignore mlock limit after perf_event_mlock_kb without CAP_IPC_LOCK\n"
 		 ">= 0: Disallow raw and ftrace function tracepoint access\n"
 		 ">= 1: Disallow CPU event access\n"
 		 ">= 2: Disallow kernel profiling\n"
+		 ">= 3: Disallow use of any event\n"
 		 "To make the adjusted perf_event_paranoid setting permanent preserve it\n"
 		 "in /etc/sysctl.conf (e.g. kernel.perf_event_paranoid = <setting>)",
 		 perf_event_paranoid());
 	case ENOENT:
 		return scnprintf(msg, size, "The %s event is not supported.", evsel__name(evsel));
-- 
2.52.0

